src/extra/vui/vui.cpp

```diff
@@ -0,0 +1,243 @@
+      }
+      releaseVariable(_vuiContextVar);
+    }
+  }
+
+  SHTypeInfo compose(SHInstanceData &data) {
+    _exposedVariables = ExposedInfo(data.shared);
+    _exposedVariables.push_back(VUIContext::VariableInfo);
+    data.shared = SHExposedTypesInfo(_exposedVariables);
+
+    _contents.compose(data);
+
+    if (_queue->valueType == SHType::None)
+      throw ComposeError("Queue is required");
+
+    if (_view->valueType == SHType::None)
+      throw ComposeError("View is required");
+
+    return shards::CoreInfo::NoneType;
+  }
+
+  SHVar activate(SHContext *shContext, const SHVar &input) {
+    auto &viewStack = _graphicsContext->renderer->getViewStack();
+    auto viewStackTop = viewStack.getOutput();
+
+    auto &queue = *varAsObjectChecked<gfx::SHDrawQueue>(_queue.get(), gfx::Types::DrawQueue);
+    auto &view = *varAsObjectChecked<gfx::SHView>(_view.get(), gfx::Types::View);
+
+    // TODO: Move to input context
+    _inputBuffer.clear();
+    for (auto &event : _inputContext->events)
+      _inputBuffer.push_back(event);
+
+    // Evaluate all UI panels
+    _vuiContext.activationContext = shContext;
+    _vuiContext.context.virtualPointScale = _scale.payload.floatValue;
+    withObjectVariable(*_vuiContextVar, &_vuiContext, VUIContext::Type, [&]() {
+      gfx::SizedView sizedView(view.view, gfx::float2(viewStackTop.viewport.getSize()));
+      _vuiContext.context.prepareInputs(_inputBuffer, _graphicsContext->window->getInputScale(), sizedView);
+      _vuiContext.context.evaluate(queue.queue, _graphicsContext->time, _graphicsContext->deltaTime);
+    });
+    _vuiContext.activationContext = nullptr;
+
+    // Render debug overlay
+    if ((bool)_debug) {
+      if (!_debugRenderer) {
+        _debugRenderer = std::make_shared<gfx::GizmoRenderer>();
+      }
+
+      _debugRenderer->begin(view.view, gfx::float2(viewStackTop.viewport.getSize()));
+      _vuiContext.context.renderDebug(_debugRenderer->getShapeRenderer());
+      _debugRenderer->end(queue.queue);
+    }
+
+    SHVar output{};
+    return output;
+  }
+};
+
+struct VUIPanelShard {
+  PARAM_PARAMVAR(_transform, "Transform", "The world transform of this panel.",
+                 {CoreInfo::Float4x4Type, Type::VariableOf(CoreInfo::Float4x4Type)});
+  PARAM_PARAMVAR(_size, "Size", "The size of the panel.", {CoreInfo::Float2Type, Type::VariableOf(CoreInfo::Float2Type)});
+  PARAM(ShardsVar, _contents, "Contents", "The panel UI contents.", {CoreInfo::ShardsOrNone});
+  PARAM_IMPL(VUIPanelShard, PARAM_IMPL_FOR(_transform), PARAM_IMPL_FOR(_size), PARAM_IMPL_FOR(_contents));
+
+  RequiredVUIContext _context;
+  Shard *_uiShard{};
+  EguiContext _eguiContext;
+  ExposedInfo _exposedTypes;
+  std::shared_ptr<Panel> _panel;
+
+  static SHTypesInfo inputTypes() { return CoreInfo::NoneType; }
+  static SHTypesInfo outputTypes() { return CoreInfo::NoneType; }
```

Comment:
Similar remark. I would expect the shard to be "pass through".

Reply:
Did passthrough

Reply:
Although `VUI.Panel` is not supposed to be used directly so input/output is not entirely accurate. I set InputType to any for now